pkg/virtctl/vmexport/vmexport.go

```diff
@@ -383,67 +483,124 @@ func downloadVolume(client kubecli.KubevirtClient, vmexport *exportv1.VirtualMac
 		return err
 	}
 
-	resp, err := handleHTTPRequest(client, vmexport, downloadUrl, vmeInfo.Insecure)
+	resp, err := HandleHTTPRequest(client, vmexport, downloadUrl, vmeInfo.Insecure, vmeInfo.ServiceURL, nil)
 	if err != nil {
 		return err
 	}
 	defer resp.Body.Close()
 
 	// Check server response
 	if resp.StatusCode != http.StatusOK {
-		return fmt.Errorf("Bad status: %s", resp.Status)
+		return fmt.Errorf("bad status: %s", resp.Status)
 	}
 
-	output, err := os.Create(vmeInfo.OutputFile)
-	if err != nil {
-		return err
-	}
-	defer util.CloseIOAndCheckErr(output, nil)
-
 	// Lastly, copy the file to the expected output
-	if err := copyFileWithProgressBar(output, resp); err != nil {
+	if err := copyFileWithProgressBar(vmeInfo.OutputWriter, resp); err != nil {
 		return err
 	}
 
-	fmt.Println("Download finished succesfully")
+	// Prevent this output ending up in the stdout
+	if vmeInfo.OutputFile != "" {
+		fmt.Println("Download finished succesfully")
+	}
 	return nil
 }
 
+func replaceUrlWithServiceUrl(manifestUrl string, vmeInfo *VMExportInfo) (string, error) {
+	// Replace internal URL with specified URL
+	manUrl, err := url.Parse(manifestUrl)
+	if err != nil {
+		return "", err
+	}
+	if vmeInfo.ServiceURL != "" {
+		manUrl.Host = vmeInfo.ServiceURL
+	}
+	return manUrl.String(), nil
+}
+
 // GetUrlFromVirtualMachineExport inspects the VirtualMachineExport status to fetch the extected URL
 func GetUrlFromVirtualMachineExport(vmexport *exportv1.VirtualMachineExport, vmeInfo *VMExportInfo) (string, error) {
-	var downloadUrl string
+	var (
+		downloadUrl string
+		err         error
+		links       *exportv1.VirtualMachineExportLink
+	)
 
-	if vmexport.Status.Links == nil || vmexport.Status.Links.External == nil || len(vmexport.Status.Links.External.Volumes) <= 0 {
-		return "", fmt.Errorf("Unable to access the volume info from '%s/%s' VirtualMachineExport", vmexport.Namespace, vmexport.Name)
+	if vmeInfo.ServiceURL == "" {
+		if vmexport.Status.Links == nil || vmexport.Status.Links.External == nil {
+			return "", fmt.Errorf("unable to access the manifest info from '%s/%s' VirtualMachineExport", vmexport.Namespace, vmexport.Name)
+		}
+		links = vmexport.Status.Links.External
+	} else {
+		if vmexport.Status.Links == nil || vmexport.Status.Links.Internal == nil {
+			return "", fmt.Errorf("unable to access the volume info from '%s/%s' VirtualMachineExport", vmexport.Namespace, vmexport.Name)
+		}
+		links = vmexport.Status.Links.Internal
 	}
-
-	volumeNumber := len(vmexport.Status.Links.External.Volumes)
+	if len(links.Volumes) <= 0 {
+		return "", fmt.Errorf("unable to access the volume info from '%s/%s' VirtualMachineExport", vmexport.Namespace, vmexport.Name)
```

Comment:
Since the same error msg is repeated three times, maybe we could clean this error handling. Something like:

```go
if vmeInfo.ServiceURL == "" && vmexport.Status.Links != nil && vmexport.Status.Links.External != nil {
    links = vmexport.Status.Links.External
} else if vmexport.Status.Links != nil && vmexport.Status.Links.Internal != nil {
    links = vmexport.Status.Links.Internal
}
if links == nil  || len(links.Volumes) <= 0 {
    return "", fmt.Errorf("unable to access the volume info from '%s/%s' VirtualMachineExport", vmexport.Namespace, vmexport.Name)
```

This is just an example though.

Reply:
I would still prioritize readability over saving a couple of lines. If you think this is harder to understand feel free to not change it.

Reply:
Fixed